dulwich/porcelain.py

```diff
@@ -191,6 +191,87 @@ class RemoteExists(Error):
     """Raised when the remote already exists."""
 
 
+def _parse_timezone_format(tz_str):
+    """Parse given string and attempt to return a timezone offset.
+    Different formats are considered in the following order:
+        - Git internal format: <unix timestamp> <timezone offset>
+        - RFC 2822: e.g. Mon, 20 Nov 1995 19:12:08 -0500
+        - ISO 8601: e.g. 1995-11-20T19:12:08-0500
+            - uses regex if dateutil is not available
+    Will default to UTC if timezone cannot be determined.
+    Args:
+      tz_str: datetime string
+    Returns: Timezone offset as integer
+    """
+    import re
+
+    # Git internal format
+    internal_format_pattern = re.compile("^[0-9]+ [+-][0-9]{,4}$")
+    if re.match(internal_format_pattern, tz_str):
+        try:
+            tz_internal = parse_timezone(tz_str.split(" ")[1].encode(DEFAULT_ENCODING))
+            return tz_internal[0]
+        except ValueError:
+            pass
+
+    # RFC 2822
+    import email.utils
+    rfc_2822 = email.utils.parsedate_tz(tz_str)
+    if rfc_2822:
+        return rfc_2822[9]
+
+    # ISO 8601
+    try:
+        from dateutil.parser import isoparse
+        try:
+            iso_8601 = isoparse(tz_str)
+            delta = iso_8601.tzinfo._offset
+            if delta:
+                # From dateutil doc:
+                # `The time zone offset in seconds, or (since version 2.6.0, represented as a datetime.timedelta object).`
+                if isinstance(delta, int):
+                    return delta
+                return int(delta.total_seconds())
+        except (AttributeError, TypeError, ValueError):
+            pass
+    except (ModuleNotFoundError, ImportError):
+        # Supported offsets:
+        # sHHMM, sHH:MM, sHH
+        iso_8601_pattern = re.compile("[0-9]([+-])([0-9]{2})(?::(?=[0-9]{2}))?([0-9]{2})?$")
```

Comment:
this is actually just a small subset of iso8601. Does git support "proper" iso8601 or just rfc3339 ?

https://stackoverflow.com/questions/522251/whats-the-difference-between-iso-8601-and-rfc-3339-date-formats

Reply:
Haven't really dug into the code of git itself.
I was mainly using [this StackOverflow answer](https://stackoverflow.com/questions/19742345/what-is-the-format-for-date-parameter-of-git-commit/19742762#19742762) and [this test](https://github.com/git/git/blob/96b2d4fa927c5055adc5b1d08f10a5d7352e2989/t/t6300-for-each-ref.sh#L128) as guidance.